lib/kubeclient/common.rb

```diff
@@ -1,255 +1,299 @@
-          # method call for get each entities
-          # build hash of entity name to array of the entities
-          entity_name = self.class.pluralize_entity entity_type.underscore
-          method_name = "get_#{entity_name}"
-          key_name = entity_type.underscore
-          result_hash[key_name] = send(method_name)
-        end
+    def create_entity(entity_type, entity_config, klass)
+      # to_hash should be called because of issue #9 in recursive open
+      # struct
+      hash = entity_config.to_hash
+
+      ns_prefix = build_namespace_prefix(entity_config.metadata['table'][:namespace])
+
+      # TODO: temporary solution to add "kind" and apiVersion to request
+      # until this issue is solved
+      # https://github.com/GoogleCloudPlatform/kubernetes/issues/6439
+      hash['kind'] = entity_type
+      hash['apiVersion'] = @api_version
+      response = handle_exception do
+        rest_client[ns_prefix + resource_name(entity_type)]
+        .post(hash.to_json, @headers)
       end
+      result = JSON.parse(response)
+      new_entity(result, klass)
+    end
 
-      def proxy_url(kind, name, port, namespace = '')
-        entity_name_plural = self.class.pluralize_entity(kind.to_s)
-        ns_prefix = build_namespace_prefix(namespace)
-        # TODO: Change this once services supports the new scheme
-        if entity_name_plural == 'pods'
-          rest_client["#{ns_prefix}#{entity_name_plural}/#{name}:#{port}/proxy"].url
-        else
-          rest_client["proxy/#{ns_prefix}#{entity_name_plural}/#{name}:#{port}"].url
-        end
+    def update_entity(entity_type, entity_config)
+      name = entity_config.metadata.name
+      # to_hash should be called because of issue #9 in recursive open
+      # struct
+      hash = entity_config.to_hash
+      ns_prefix = build_namespace_prefix(entity_config.metadata['table'][:namespace])
+      handle_exception do
+        rest_client[ns_prefix + resource_name(entity_type) + "/#{name}"]
+          .put(hash.to_json, @headers)
       end
+    end
 
-      def resource_name(entity_type)
-        self.class.pluralize_entity entity_type.downcase
-      end
+    def new_entity(hash, klass)
+      klass.new(hash)
+    end
 
-      def api_valid?
-        result = api
-        result.is_a?(Hash) && (result['versions'] || []).include?(@api_version)
+    def retrieve_all_entities(entity_types)
+      entity_types.each_with_object({}) do |(_, entity_type), result_hash|
+        # method call for get each entities
+        # build hash of entity name to array of the entities
+        entity_name = ClientMixin.pluralize_entity entity_type.underscore
+        method_name = "get_#{entity_name}"
+        key_name = entity_type.underscore
+        result_hash[key_name] = send(method_name)
       end
+    end
 
-      def api
-        response = handle_exception do
-          create_rest_client.get(@headers)
-        end
-        JSON.parse(response)
+    def resource_name(entity_type)
```

Comment:
@moolitayer you mistakenly inverted the order of the two methods `resource_name` and `proxy_url`

`resource_name` was after `proxy_url`